HIP/hip-796.md

```diff
@@ -0,0 +1,728 @@
+
+## Example
+
+The concept of partitioning and locking can be illustrated using an example. Consider a `token-definition` for `AcmeToken` that represents investments in a fund.
+
+- The token issuer creates this `token-definition` with a `partition-key`. Let’s say this has a `TokenID` of `0.0.123456`.
+- The fund has different tranches with three different timelines until which these funds are locked (i.e. cannot be transferred or traded by a holder). Let’s call these tranches Tranche-A (locked until Jan 1, 2024), Tranche-B (locked until Jan 1, 2025), and Tranche-C (locked until Jan 1, 2026).
+- The token issuer creates three `partition-definition`s for the token-definition `0.0.123456`. Hedera creates these three definitions and returns `TokenID`s of `0.0.200001`, `0.0.200002`, and `0.0.200003` that represent AcmeToken.Tranche-A, AcmeToken.Tranche-B, AcmeToken.Tranche-C respectively.
+- The token issuer sets three different metadata on these partition definitions independently of each other.
+- When anybody queries the mirror node for this token id `0.0.123456`, they get an equivalent of:
+
+```
+{
+    id: 0.0.123456,
+    name: AcmeToken,
+    symbol: ACME,
+    lockKey: 0x111abc..,
+    partitionKey: 0x222abc..,
+    partitionMoveKey: 0x333ace…,
+    partitions: [
+        {partition_number:1, metadata: “ipfs://xxx”, partition_id: 0.0.200001},
+        {partition_number:2, metadata: “ipfs://yyy”, partition_id: 0.0.200002},
+        {partition_number:3, metadata: “ipfs://zzz”, partition_id: 0.0.200003}
+    ]
+}
+```
+
+- Alice makes investments in this fund in two of these tranches - Tranche-A and Tranche-C. She skips Tranche-B.
+- A query to the mirror node for her account balance shows the equivalent of:
+
+```
+{
+…
+tokens: {[
+    {
+        token_id: 0.0.123456,
+        balance: [
+            {partition_id: 1, partition_id: 0.0.200001, balance: 10, locked: 5 },
+            {partition_id: 3, partition_id: 0.0.200003, balance: 10, locked: 7 },
+        ]
+    }
+]}
+…
+}
+```
+
+- On 1 June 2024 the holder of the `lock-key` signs a transaction to unlock the tokens in Tranche-A.
+- Alice can attempt to transfer tokens in an individual partition on 1 June 2024.
+- If she transfers tokens from partition-1, that transfer will succeed.
+- If she transfers tokens from partition-3, that transfer will return an appropriate error informing her that her tokens are locked.
+- To make the transfer, she will use the simple form of the current crypto transfer as the existing tokenTransferList. If she wants to transfer tokens from partition-3, she (or the SDK) queries the mirror nodes or the wallets and maps partition-3 to the TokenID of `0.0.200003` and fills that in the token transfer list of the cryptoTransfer transaction.
+
+## Specification
```

Comment:
This proposed HIP has a lot of moving parts and though the wording seems clear I can’t help but think it would be more solid if it included some diagrams and/or pseudocode to make things especially concrete.  There’s a lot of room for implementation fuzziness to sneak in to a big pile of words.  At least if there were diagrams and/or pseudocode a contradiction between the two could be noticed and addressed - it might be either the words or the diagram/code - whereas as it stands things might slip though the implementation.  And not just contradictions, but also holes.

In the same way, additional tables would be useful, e.g., which combination of parameters to APIs are _valid_ and which are not?  E.g., some of these APIs have 4 "buckets" on 2-dimensions:  fungible vs non-fungible, and token-definition-id vs partition-id.  It's not clear that _all_ parameters of some of these APIs are acceptable to all 4 buckets.  (And I'm _not_ talking about the difference between "amount" and "serial#", that's clear enough.)